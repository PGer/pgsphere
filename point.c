#include "point.h"

/*!        
  \file
  This file contains definitions for spherical
  point functions
  \brief definitions of spherical point functions
*/ 


#ifndef DOXYGEN_SHOULD_SKIP_THIS

 PG_FUNCTION_INFO_V1(spherepoint_in);
 PG_FUNCTION_INFO_V1(spherepoint_from_long_lat);
 PG_FUNCTION_INFO_V1(spherepoint_distance);
 PG_FUNCTION_INFO_V1(spherepoint_long);
 PG_FUNCTION_INFO_V1(spherepoint_lat);
 PG_FUNCTION_INFO_V1(spherepoint_x);
 PG_FUNCTION_INFO_V1(spherepoint_y);
 PG_FUNCTION_INFO_V1(spherepoint_z);
 PG_FUNCTION_INFO_V1(spherepoint_xyz);
 PG_FUNCTION_INFO_V1(spherepoint_equal);

#endif

  bool spoint_eq ( const SPoint * p1 , const SPoint * p2 )
  {
    Vector3D a,b ;
    spoint_vector3d ( &a , p1 );
    spoint_vector3d ( &b , p2 );
    return ( vector3d_eq ( &a, &b ) );
  }


  SPoint *  spoint_check (SPoint * spoint)
  {
    static bool lat_is_neg;
    lat_is_neg = ( spoint->lat < 0 )?( TRUE ):( FALSE );

    spoint->lng  = spoint->lng - floor(spoint->lng / ( PID ) )  * PID;
    spoint->lat  = spoint->lat - floor(spoint->lat / ( PID ) )  * PID;
    if( spoint->lng < 0.0 ){
      spoint->lng +=  ( PID );
    }
    if( spoint->lat > PI   ){
      spoint->lat -=  ( 2*PI );
    }
    if( spoint->lat > PIH  ){
      spoint->lat  =  ( PI - spoint->lat );
      spoint->lng +=  ( ( spoint->lng<PI )?( PI ):( -PI ) ) ;
    }
    if( spoint->lat < -PIH ){
      spoint->lat  =  ( -PI - spoint->lat  );
      spoint->lng +=  ( ( spoint->lng<PI )?( PI ):( -PI ) ) ;
    }
    if (
        FPeq(spoint->lat,PIH) && lat_is_neg 
      ) spoint->lat  = -PIH ;
    
    if ( FPeq(spoint->lng,PID) ){
        spoint->lng  = 0.0 ;
    }
    if ( FPzero(spoint->lng) ){
        spoint->lng  = 0.0 ;
    }
    if ( FPzero(spoint->lat) ){
        spoint->lat  = 0.0 ;
    }
    
    return spoint;
  }


  SPoint * vector3d_spoint ( SPoint * p , const Vector3D * v )
  {

    double rho     =  sqrt ( (v->x)*(v->x) + (v->y)*(v->y) );
    if( 0.0 == rho ){
      if( FPzero( v->z ) ) {
        p->lat  = 0.0;
      } else if( v->z > 0 ){
        p->lat  = PIH;
      } else if( v->z < 0 ) {
        p->lat  = - PIH;
      } 
    } else {
      p->lat  = atan( v->z / rho );
    }

    p->lng   = atan2( v->y, v->x );
    if( FPzero(p->lng) ){
    	p->lng = 0.0;
    } else
    if( p->lng < 0.0 ){
    	p->lng += PID ;
    }
    return ( p );
  } 

  Vector3D * spoint_vector3d ( Vector3D * v , const SPoint * p )
  {
    v->x      = cos(p->lng)*cos(p->lat);
    v->y      = sin(p->lng)*cos(p->lat);
    v->z      = sin(p->lat);
    return ( v ) ;
  }

  Datum  spherepoint_in(PG_FUNCTION_ARGS)
  {
    SPoint  * sp  = ( SPoint * ) MALLOC ( sizeof ( SPoint ) ) ;
    char    *  c  = PG_GETARG_CSTRING(0);
    double lng, lat;

    void sphere_yyparse( void );
    init_buffer ( c );
    sphere_yyparse();
    if ( get_point( &lng, &lat ) ){
      sp->lng       = lng;
      sp->lat       = lat;
      spoint_check ( sp );
    } else {
      reset_buffer();
      FREE( sp );
      sp = NULL;
      elog ( ERROR , "spherepoint_in: parse error" );
    }
    reset_buffer();
    PG_RETURN_POINTER( sp );
  }


  Datum  spherepoint_from_long_lat(PG_FUNCTION_ARGS)
  {
    SPoint   * p  = ( SPoint * ) MALLOC ( sizeof ( SPoint ) ) ;
    p->lng        = PG_GETARG_FLOAT8( 0 );
    p->lat        = PG_GETARG_FLOAT8( 1 );
    spoint_check ( p ) ;
    PG_RETURN_POINTER( p );
  }

  float8 spoint_dist ( const SPoint * p1, const SPoint * p2 )
  {
	float8 dl = p1->lng - p2->lng;
    float8 f = (  ( sin( p1->lat )*sin( p2->lat ) + cos( p1->lat )*cos( p2->lat )*cos( dl ) ) );
    if( FPeq( f, 1.0 ) ){
   	  /* for small distances */
      Vector3D v1, v2, v3;
      spoint_vector3d(&v1, p1);
      spoint_vector3d(&v2, p2);
      vector3d_cross( &v3, &v1, &v2 );
      f = vector3d_length(&v3);
    } else {
      f = acos(f);
    }
    if ( FPzero(f) ){
      return 0.0;
    } else {
      return f;
    }
  }

  Datum spherepoint_distance(PG_FUNCTION_ARGS) 
  {

    SPoint  * p1 =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    SPoint  * p2 =  ( SPoint * ) PG_GETARG_POINTER ( 1 ) ;
    PG_RETURN_FLOAT8 ( spoint_dist ( p1, p2 ) );

  }

  Datum  spherepoint_long(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * )  PG_GETARG_POINTER ( 0 ) ;
    PG_RETURN_FLOAT8 ( p->lng ); 
  }

  Datum  spherepoint_lat(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    PG_RETURN_FLOAT8 ( p->lat ); 
  }

  Datum  spherepoint_x(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    static Vector3D  v ;
    spoint_vector3d ( &v , p );    
    PG_RETURN_FLOAT8 ( v.x   ); 
  }
  
  Datum  spherepoint_y(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    static Vector3D  v ;
    spoint_vector3d ( &v , p );    
    PG_RETURN_FLOAT8 ( v.y   ); 
  }

  Datum  spherepoint_z(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    static Vector3D  v ;
    spoint_vector3d ( &v , p );    
    PG_RETURN_FLOAT8 ( v.z   ); 
  }

  Datum  spherepoint_xyz(PG_FUNCTION_ARGS)
  {
    SPoint  * p =  ( SPoint * ) PG_GETARG_POINTER ( 0 ) ;
    Datum    dret[3];
    ArrayType  *result;
    static Vector3D  v ;
    spoint_vector3d ( &v , p );    
    dret[0] = Float8GetDatumFast(v.x);
    dret[1] = Float8GetDatumFast(v.y);
    dret[2] = Float8GetDatumFast(v.z);
	#ifdef FLOAT8PASSBYVAL
		result = construct_array ( dret , 3, FLOAT8OID, sizeof(float8), FLOAT8PASSBYVAL, 'd');
	#else
        result = construct_array ( dret , 3, FLOAT8OID, sizeof(float8), false /* float8 byval */ , 'd' );
	#endif
    PG_RETURN_ARRAYTYPE_P(result);                                                                                                          
  }

  Datum  spherepoint_equal(PG_FUNCTION_ARGS)
  {
    SPoint  * p1 =  ( SPoint * )  PG_GETARG_POINTER ( 0 ) ;
    SPoint  * p2 =  ( SPoint * )  PG_GETARG_POINTER ( 1 ) ;
    PG_RETURN_BOOL( spoint_eq ( p1 , p2 ) );
  }

