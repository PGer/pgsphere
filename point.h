#ifndef __PGS_POINT_H__
#define __PGS_POINT_H__

#include "vector3d.h"
#include "sbuffer.h"
/*!
  \file
  This file contains declaration for spherical
  point and functions
  \brief spherical point declarations
*/

/*!
  The definition of spherical point.
  \brief Spherical point
*/
  typedef struct {
   float8 lng;  //!< longitude value in radians
   float8 lat;  //!< latitude value in radians
  }  SPoint;

  /*!
    Calculate the distance of two spherical points
    \brief Distance of two spherical points
    \return distance in radians
  */
  float8 spoint_dist ( const SPoint * p1, const SPoint * p2 );


  /*!
    \brief check whether two points are equal
    \param p1 first point
    \param p2 second point
    \return true, if equal
  */
  bool spoint_eq ( const SPoint * p1 , const SPoint * p2 );

  /*!
    Checks the longitude and latitude values and take sure
    the right value ranges
    \brief  validate the spherical point
    \param  spoint pointer to spherical point
    \return pointer to spherical point
  */
  SPoint *  spoint_check (SPoint * spoint);

  /*!
    \brief transforms a 3 dim.vector to a spherical point 
    \param p pointer to spherical point
    \param v pointer to 3 dim. vector
    \return pointer to spherical point
  */
  SPoint   * vector3d_spoint    ( SPoint   * p   , const Vector3D * v );

  /*!
    \brief transforms a spherical point to a 3 dim.vector
    \param v pointer to 3 dim. vector
    \param p pointer to spherical point
    \return pointer to 3 dim. vector
  */
  Vector3D * spoint_vector3d    ( Vector3D * v   , const SPoint   * p );


  /*!
    Take the input and stores it as a spherical point
    \return a spherical point datum
    \note Does check the input too.
   */
  Datum  spherepoint_in(PG_FUNCTION_ARGS);


  /*!
    Create a spherical point from longitude
    and latitude both in radians
    \brief  point created from longitude and latitude
    \return a spherical point datum
    \note PostgreSQL function
   */
  Datum  spherepoint_from_long_lat(PG_FUNCTION_ARGS);


  /*!
    Calculate the distance of two spherical points
    \brief Distance of two spherical points
    \return distance ( float8 ) as datum
    \see spoint_dist ( SPoint *, SPoint * )
    \note PostgreSQL function
  */
  Datum spherepoint_distance(PG_FUNCTION_ARGS);

  /*!
    \brief longitude of spherical point
    \return longitude datum in radians ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_long(PG_FUNCTION_ARGS);

  /*!
    \brief latitude of spherical point
    \return latitude datum in radians ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_lat(PG_FUNCTION_ARGS);

  /*!
    \brief cartesian x-value of spherical point
    \return x-value datum ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_x(PG_FUNCTION_ARGS);

  /*!
    \brief cartesian y-value of spherical point
    \return y-value datum ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_y(PG_FUNCTION_ARGS);

  /*!
    \brief cartesian z-value of spherical point
    \return z-value datum ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_z(PG_FUNCTION_ARGS);

  /*!
    \brief cartesian values of spherical point
    \return array datum ( float8 )
    \note PostgreSQL function
  */
  Datum  spherepoint_xyz(PG_FUNCTION_ARGS);

  /*!
    \brief check whether two points are equal
    \return boolean datum
    \note PostgreSQL function
  */
  Datum  spherepoint_equal(PG_FUNCTION_ARGS);


#endif
