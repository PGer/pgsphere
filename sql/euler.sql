-- checking Euler transformation operators

SELECT strans '-10d,0d,10d,ZZZ'  = '-10d,0d,10d,XXX' ;
SELECT strans '-40d,0d,40d,ZZZ' <> '-40d,0d,40d,XXX' ;

