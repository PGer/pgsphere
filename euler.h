#ifndef __PGS_EULER_H__
#define __PGS_EULER_H__

#include "point.h"

/*!
  \file
  \brief Euler transformation declarations
*/
    
  /*!
    \brief Spherical (Euler) transformation.
  */
  typedef struct {
    unsigned char   phi_a:2 , //!< first axis
                  theta_a:2 , //!< second axis
                    psi_a:2 ; //!< third axis
    float8              phi , //!< first rotation angle
                      theta , //!< second rotation angle
                        psi ; //!< third rotation angle
  } SEuler;


  /*!
    \brief  transforms a spherical point
    \param  out pointer to transformed spherical point
    \param  in  pointer to spherical input point
    \param  se  Euler transformation
    \return pointer to transformed spherical point
  */
  SPoint    * euler_spoint_trans        ( SPoint * out , const SPoint  * in , const SEuler * se );


  /*!
    \brief transforms a spherical vector to a Euler transformation
    \param spb pointer to begin of spherical vector
    \param spe pointer to end   of spherical vector
    \param se  pointer to Euler transformation
    \return true, if calculation was successfull
    \see spherevector_to_euler_inv ( SEuler *, SPoint *, SPoint * )
  */
  bool   spherevector_to_euler     ( SEuler * se, const SPoint * spb , const SPoint * spe );

  /*!
    \brief sets the axes of transformation to ZXZ
    \param se pointer to Euler transformation
  */
  void        seuler_set_zxz            ( SEuler * se );

  /*!
   \brief checks equality of two transformations
   \param e1 pointer to first transformation
   \param e2 pointer to second transformation
   \return true if equal
  */
  bool        strans_eq                 ( const SEuler * e1 , const SEuler * e2 );

  /*!
    \brief Transforms a vector using an Euler transformation
    \param se pointer to Euler transformation
    \param in pointer to input vector
    \param out pointer to result vector
    \return pointer to result vector
  */
  Vector3D  * euler_vector_trans ( Vector3D * out , const Vector3D  * in , const SEuler * se );

  /*!
    \brief inverts an Euler transformation
    \param se_in pointer to input transformation
    \param se_out pointer to inverted transformation
    \return pointer to inverted transformation
    \see  spheretrans_inv ( SEuler * );

  */
  SEuler    * spheretrans_inverse       ( SEuler * se_out , const SEuler * se_in );

  /*!
    \brief inverts an Euler transformation
    \param se pointer to input and output transformation
    \return pointer to inverted transformation
    \see spheretrans_inverse ( SEuler * , const SEuler * );

  */
  SEuler    * spheretrans_inv           ( SEuler * se );

  /*!
    \brief Converts an Euler transformation to a ZXZ-axis transformation
    \param se pointer to input transformation
    \param ret pointer to result transformation
    \return pointer to converted transformation
  */
  SEuler    * strans_zxz                ( SEuler * ret , const SEuler * se );

  /*!
    This function enables a transformation of an Euler transformation.
    The result is allways a ZXZ-axis transformation
    \brief Transforms an Euler transformation
    \param se  pointer to a transformation
    \param in  pointer to input transformation
    \param out pointer to result transformation
    \return pointer to transformed transformation
  */
  SEuler    * seuler_trans_zxz          ( SEuler * out , const SEuler * in , const SEuler * se );


  /*!
    \brief Input of Euler transformation
    \return a Euler transformation datum
    \note PostgreSQL function
  */
  Datum  spheretrans_in               (PG_FUNCTION_ARGS);

  /*!
    Input of a Euler transformation with axis Z,X,Z from three angles
    (phi, theta, psi) in radians
    \brief Input of Euler ZXZ transformation from three angles
    \return a Euler transformation datum
    \note PostgreSQL function
  */
  Datum  spheretrans_from_float8      (PG_FUNCTION_ARGS);

  /*!
    \brief Returns the first angle of Euler transformation in radians
    \return a float8 datum
    \note PostgreSQL function
  */
  Datum  spheretrans_phi              (PG_FUNCTION_ARGS);

  /*!
    \brief Returns the second angle of Euler transformation in radians
    \return a float8 datum
    \note PostgreSQL function
  */
  Datum  spheretrans_theta            (PG_FUNCTION_ARGS);

  /*!
    \brief Returns the third angle of Euler transformation in radians
    \return a float8 datum
    \note PostgreSQL function
  */
  Datum  spheretrans_psi              (PG_FUNCTION_ARGS);

  /*!
    \brief Returns the axis of Euler transformation as three letter code
    \return a character varying datum
    \note PostgreSQL function
  */
  Datum  spheretrans_type             (PG_FUNCTION_ARGS);

  /*!
    This function is needed for +strans operator
    \brief Returns the Euler transformation ( does nothing )
    \return Euler transformation datum
    \note PostgreSQL function
  */
  Datum  spheretrans                  (PG_FUNCTION_ARGS);

  /*!
    \brief Returns the inverse Euler transformation
    \return Euler transformation datum
    \note PostgreSQL function
  */
  Datum  spheretrans_invert           (PG_FUNCTION_ARGS);

  /*!
    \brief Convert an Euler transformation to a ZXZ-axis transformation
    \return Euler transformation datum
    \note PostgreSQL function
  */
  Datum  spheretrans_zxz              (PG_FUNCTION_ARGS);

  /*!
    This function creates a Euler transformation from 3 angle values
    in radians and three letter code used for axes. A letter can be
    X,Y or Z ( not case sensitive ).
    \brief Creates an Euler transformation from angles and axes
    \return Euler transformation datum
    \note PostgreSQL function
  */
  Datum  spheretrans_from_float8_and_type (PG_FUNCTION_ARGS);


  /*!
    \brief Checks equality of two Euler transformations
    \return bool datum, true if equal
    \note PostgreSQL function
  */
  Datum  spheretrans_equal            (PG_FUNCTION_ARGS);

  /*!
    \brief Checks inequality of two Euler transformations
    \return bool datum, true if not equal
    \note PostgreSQL function
  */
  Datum  spheretrans_not_equal        (PG_FUNCTION_ARGS);


  /*!
    \brief Transforms an Euler transformations
    \return Euler transformation datum
    \note PostgreSQL function
  */
  Datum  spheretrans_trans            (PG_FUNCTION_ARGS);

  /*!
    \brief Transforms inverse an Euler transformations
    \return Euler transformation datum
    \note PostgreSQL function
  */
  Datum  spheretrans_trans_inv        (PG_FUNCTION_ARGS);

  /*!
    \brief transforms a spherical point
    \return datum of transformed spherical point
    \note PostgreSQL function
  */
  Datum  spheretrans_point            (PG_FUNCTION_ARGS);

  /*!
    \brief transforms inverse a spherical point
    \return datum of inverse transformed spherical point
    \note PostgreSQL function
  */
  Datum  spheretrans_point_inverse    (PG_FUNCTION_ARGS);


#endif
