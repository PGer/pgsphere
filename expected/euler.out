-- checking Euler transformation operators
SELECT strans '-10d,0d,10d,ZZZ'  = '-10d,0d,10d,XXX' ;
 ?column? 
----------
 t
(1 row)

SELECT strans '-40d,0d,40d,ZZZ' <> '-40d,0d,40d,XXX' ;
 ?column? 
----------
 f
(1 row)

